### PR TITLE
[Enhancement][Refactor](Nereids) generate pattern by operator and refactor Plan and NODE_TYPE generic type

### PR DESCRIPTION
**Pull request by 924060929**
# Proposed changes

This pr support
1. remove the generic type from operator, remove some NODE_TYPE from plan and expression
2. refactor Plan and NODE_TYPE generic type
3. support child class matching by TypePattern
4. analyze the code of operator and generate pattern makes it easy to create rules.

e.g. 
```java
class LogicalJoin extends LogicalBinaryOperator;
class PhysicalFilter extends PhysicalUnaryOperator;
```

will generate the code
```java
interface GeneratedPatterns extends Patterns {
  default PatternDescriptor<LogicalBinaryPlan<LogicalJoin, Plan, Plan>, Plan> logicalJoin() {
      return new PatternDescriptor<LogicalBinaryPlan<LogicalJoin, Plan, Plan>, Plan>(
          new TypePattern(LogicalJoin.class, Pattern.FIXED, Pattern.FIXED),
          defaultPromise()
      );
  }
  
  default <C1 extends Plan, C2 extends Plan>
  PatternDescriptor<LogicalBinaryPlan<LogicalJoin, C1, C2>, Plan>
          logicalJoin(PatternDescriptor<C1, Plan> child1, PatternDescriptor<C2, Plan> child2) {
      return new PatternDescriptor<LogicalBinaryPlan<LogicalJoin, C1, C2>, Plan>(
          new TypePattern(LogicalJoin.class, child1.pattern, child2.pattern),
          defaultPromise()
      );
  }

  default PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, Plan>, Plan> physicalFilter() {
      return new PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, Plan>, Plan>(
          new TypePattern(PhysicalFilter.class, Pattern.FIXED),
          defaultPromise()
      );
  }
  
  default <C1 extends Plan>
  PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, C1>, Plan>
          physicalFilter(PatternDescriptor<C1, Plan> child1) {
      return new PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, C1>, Plan>(
          new TypePattern(PhysicalFilter.class, child1.pattern),
          defaultPromise()
      );
  }
}
```
and then we don't have to add pattern for new operators.

this function utilizing jsr269 to do something in compile time, and utilizing antlr4 to analyze the code of `Operator`, then we can generate corresponding pattern.

pattern generate steps:
1. maven-compiler-plugin in the pom.xml will compile fe-core three terms. first term will compile `PatternDescribable.java` and `PatternDescribableProcessor.java`
2. second compile term will compile `PatternDescribableProcessPoint.java`, and enable annotation process `PatternDescribableProcessor`, PatternDescribableProcessor will receive the event and know that `PatternDescribableProcessPoint` class contains the `PatternDescribable` annotation.
3. `PatternDescribableProcessor` will not process `PatternDescribableProcessPoint`, but find all java file exists in `operatorPath` that specify in pom.xml, and then parse to Java AST(abstract syntax tree).
5. PatternDescribableProcessor collect java AST and use `PatternGeneratorAnalyzer` to analyze AST, find the child class file for `PlanOperator` then generate `GeneratedPatterns.java` by the AST.
6. third compile term will compile `GeneratedPatterns.java` and other java file.

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: No
7. Has unit tests been added: No Need
8. Has document been added or modified: No Need
9. Does it need to update dependencies: No
10. Are there any changes that cannot be rolled back: Yes

### CODE REVIEW

**Comment by wangshuo128** [comment]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java [hunk 1]
```diff
@@ -42,6 +42,7 @@ public enum OperatorType {
     PHYSICAL_BROADCAST_HASH_JOIN,
 
     // pattern
+    NORMAL_PATTERN,
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/operators/OperatorType.java:45]: Pattern must has an `OperatorType`, so add the `NORMAL_PATTERN` distinct with ANY / FIXED

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaParser.g4 [hunk 2]
```diff
@@ -0,0 +1,781 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Copied from Antlr and modified for Apache Doris
+
+/*
```
Comment (morningman) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaParser.g4:20]: Not modify its origin license

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaLexer.g4 [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Copied from Antlr and modified for Apache Doris
+
+/*
```
[the remark this answers is not in the document]
Comment (924060929) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaLexer.g4:1]: done

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaParser.g4 (hunk 2 above)
Comment (924060929) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/JavaParser.g4:1]: done

**Review by morrySnow — DISMISSED** [review]

**Review by EmmyMiao87 — COMMENTED** [review]
LGTM

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**